### PR TITLE
model: Support Hybrid Mamba2 NemotronHForCausalLM (nvidia/NVIDIA-Nemotron-Nano-9B-v2)

### PR DESCRIPTION
**Pull request by netanel-haber**
Support the `NemotronHForCausalLM` architecture, which can include any combination of *Mamba2*, MLP and normal self-attention layers.

## Motivation
Support the `NemotronHForCausalLM` architecture, which includes https://huggingface.co/nvidia/NVIDIA-Nemotron-Nano-9B-v2:
> The model uses a hybrid architecture consisting primarily of Mamba-2 and MLP layers combined with just four Attention layers. For the architecture, please refer to the [Nemotron-H tech report](https://arxiv.org/abs/2504.03624).

TODOs:
- [x] Measure runtime of mamba kernel unittests and update run_test.
- [x] Run a GSM8K to check the Qwen3-Next + Falcon haven't been impacted by the changes.
- [x] Perhaps get official copy for Nemotron-Nano-V2 support matrix entry.

---------------------------
## Checklist
- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: test/srt/run_suite.py [hunk 1]
```diff
@@ -123,6 +125,9 @@ class TestFile:
         TestFile("test_vlm_input_format.py", 300),
         TestFile("test_vision_openai_server_a.py", 724),
         TestFile("test_vision_openai_server_b.py", 446),
+        TestFile("layers/attention/mamba/test_causal_conv1d.py", TODO),  # TODO: Measure
```
Comment (yizhang2077) [inline test/srt/run_suite.py]: add time here?

Comment (ispobock) [inline test/srt/run_suite.py]: Could you update the estimated time?

File: python/sglang/srt/layers/attention/mamba/mamba2_mixer.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+from .mamba2_metadata import Mamba2Metadata
+from .mamba2_mixer_rms_norm_gated import Mixer2RMSNormGated
+from .mamba_ssm import selective_state_update
+from .ssd_combined import mamba_chunk_scan_combined
+
+# Added by the IBM Team, 2024
+
+
+# Adapted from transformers.models.mamba.modeling_mamba.MambaMixer
+class Mamba2Mixer(nn.Module):
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/mamba/mamba2_mixer.py]: Some mamba components are already added in https://github.com/sgl-project/sglang/pull/10988. Could you rebase main branch and remove the common parts?

**Comment by ispobock** [comment]
@netanel-haber Could you also resolve the conflicts? Thanks!

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -107,17 +110,39 @@ def clear(self):
...
+        cache_params: Mamba2CacheParams,
         device: str,
         speculative_num_draft_tokens: Optional[int] = None,
     ):
+        conv_state_shape = cache_params.shape.conv
+        temporal_state_shape = cache_params.shape.temporal
+        conv_dtype = cache_params.dtype.conv
+        ssm_dtype = cache_params.dtype.temporal
+        num_mamba_layers = len(cache_params.layers)
+        # TODO: why +1?
```
Comment (mickqian) [inline python/sglang/srt/mem_cache/memory_pool.py]: is this still necessary?

File: python/sglang/srt/layers/attention/mamba/mamba2_mixer.py (hunk 2 above)
Comment (netanel-haber) [reply python/sglang/srt/layers/attention/mamba/mamba2_mixer.py]: Done. I merged the two Mamba2Mixers, and reused kernels added to .ops by falcon merge. PR has therefore been reduced from 6k LOC to ~3K LOC (1K of which is unittests lifted almost wholesale from VLLM).
I'll actually run the merge result on Sunday and change minor things if anything broke that I hadn't noticed, and go over my TODOs in my pr description, including the unittest times.
I added a TODO to the PR description to check Falcon hasn't been damaged either.

File: test/srt/run_suite.py (hunk 1 above)
Comment (netanel-haber) [reply test/srt/run_suite.py]: See: https://github.com/sgl-project/sglang/pull/10909#discussion_r2401580981

File: python/sglang/srt/layers/attention/mamba/ops/ssd_state_passing.py [hunk 4]
```diff
@@ -13,17 +13,18 @@
...
-#         triton.Config({"BLOCK_SIZE": 64}),
-#         triton.Config({"BLOCK_SIZE": 128}),
-#         triton.Config({"BLOCK_SIZE": 256}),
-#         triton.Config({"BLOCK_SIZE": 512}),
-#         triton.Config({"BLOCK_SIZE": 1024}),
-#         triton.Config({"BLOCK_SIZE": 2048}),
-#     ],
-#     key=["dim"],
-# )
+@triton.autotune(
```
Comment (yizhang2077) [inline python/sglang/srt/layers/attention/mamba/ops/ssd_state_passing.py]: remove autotune here, the same as other kernel

File: test/srt/run_suite.py [hunk 5]
```diff
@@ -11,6 +11,8 @@ class TestFile:
     estimated_time: float = 60
 
 
+TODO = 500
```
Comment (yizhang2077) [inline test/srt/run_suite.py]: remove TODO here

File: python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py [hunk 6]
```diff
@@ -672,7 +672,9 @@ def _causal_conv1d_update_kernel(
         + (conv_state_batch_coord * stride_conv_state_seq)
         + conv_state_token_offset * stride_conv_state_tok
         + (idx_feats * stride_conv_state_dim)[None, :]
-        + ((idx_tokens + 1) * stride_conv_state_tok)[:, None]
+        + ((idx_tokens + (1 if IS_SPEC_DECODING else seqlen)) * stride_conv_state_tok)[
+            :, None
```
Comment (yizhang2077) [inline python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py:676]: why do we need modify triton kernel here? cc @hebiao064

Comment (netanel-haber) [reply python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py:676]: https://github.com/vllm-project/vllm/pull/24680
As per fix introduced to VLLM - for non spec decoding models (like nanov2)

File: python/sglang/srt/layers/attention/mamba/ops/ssd_state_passing.py (hunk 4 above)
Comment (netanel-haber) [reply python/sglang/srt/layers/attention/mamba/ops/ssd_state_passing.py]: Why remove the autotune?

Comment (yizhang2077) [reply python/sglang/srt/layers/attention/mamba/ops/ssd_state_passing.py]: autotune will cause hang in profile stage. I think the best approach is to add a config, and finetune offline. Then we add it this config into a folder. We do `fusedMoE` in this way. But if we do in this way there are so many config/environments  to tune and it is hard to maintain. So I think we can choose an any proper config if the triton ops is not heavy as a trade-off.

File: python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py (hunk 6 above)
Comment (yizhang2077) [reply python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py:676]: I think we need ask @hebiao064

**Comment by ispobock** [comment]
@netanel-haber could you resolve the conflicts?

File: python/sglang/srt/layers/attention/mamba/ops/ssd_state_passing.py (hunk 4 above)
Comment (netanel-haber) [reply python/sglang/srt/layers/attention/mamba/ops/ssd_state_passing.py]: Removed autotune code.

File: python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py (hunk 6 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py:676]: @netanel-haber tho we adapted the kernel from vllm (which is also adapted from FLA repo), SGLang is tacking the Spec Decoding differently than vllm.

In SGLang, IS_SPEC_DECODING will always be False even under Target Verify mode, we allocated a conv states pool with 4 (as the qwen3-next conv state setup) + spec_num_steps (typically 3 or 4), and after verification is done, we sliced the 4 + x (x = num_accepted tokens) and put it back to the pool.

So we never used the spec decoding related logic in the triton kernel, I believe your PR and QWen3-Next will be good with or without this change.


I tentatively agree with keep your change to keep synced with vLLM, but please try test spec decoding with your model and also QWen3-Next

Comment (netanel-haber) [reply python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py:676]: I checked, and `IS_SPEC_DECODING` is always `False` for Qwen3-Next, even if speculative decoding is running, as you said.
Anyway, testing with or without the change yields equivalent results for Qwen3-Next with both spec decoding on and off [and for NanoV2, without spec decoding].

```python
test/srt/models/test_qwen3_next_models.py::TestQwen3Next::test_gsm8k
metrics={'accuracy': np.float64(0.965)}

test/srt/models/test_qwen3_next_models.py::TestQwen3NextMTP::test_gsm8k
metrics={'accuracy': np.float64(0.96)}

test/srt/models/test_nvidia_nemotron_nano_v2.py
metrics={'accuracy': np.float64(0.885)}

…/sglang git checkout main -- python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py

test/srt/models/test_qwen3_next_models.py::TestQwen3Next::test_gsm8k
metrics={'accuracy': np.float64(0.94)}

test/srt/models/test_qwen3_next_models.py::TestQwen3NextMTP::test_gsm8k
metrics={'accuracy': np.float64(0.965)}

test/srt/models/test_nvidia_nemotron_nano_v2.py
metrics={'accuracy': np.float64(0.875)}
```


So, I think that since you agree that parity with VLLM has merit, and that we observe no accuracy change, I'm leaning towards keeping the change.
I'll wait for further comment from you here before I resolve

**Review by ispobock — APPROVED** [review]
Overall LGTM. @netanel-haber just need to resolve the conflict and get @yizhang2077 @hebiao064 's approval. Then it can be merged.

File: python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py (hunk 6 above)
Comment (netanel-haber) [reply python/sglang/srt/layers/attention/mamba/causal_conv1d_triton.py:676]: Resolving at this point, if @hebiao064 wants to, he can unresolve

**Review by yizhang2077 — APPROVED** [review]
LGTM overall, once all ut about linear models have passed I think we could merge it

File: test/srt/run_suite.py [hunk 7]
```diff
@@ -127,6 +127,10 @@ class TestFile:
         TestFile("test_vlm_input_format.py", 300),
         TestFile("test_vision_openai_server_a.py", 724),
         TestFile("test_vision_openai_server_b.py", 446),
+        TestFile("layers/attention/mamba/test_causal_conv1d.py", 85),
+        TestFile("layers/attention/mamba/test_mamba_ssm.py", 85),
+        TestFile("layers/attention/mamba/test_mamba_ssm_ssd.py", 220),
+        TestFile("models/test_nvidia_nemotron_nano_v2.py", 180),
```
Comment (merrymercy) [inline test/srt/run_suite.py:133]: 1. These tests won't run by CI. You need to add this line https://github.com/sgl-project/sglang/blob/55b14656e656c0a0be3d9a24e69f624c957429e7/python/sglang/srt/mem_cache/storage/hf3fs/test_hf3fs_utils.py#L43
2. If some layer-wise unit tests take a long time to run in CI, we can consider skipping them in our per-commit CI (but still save the files in the repo for your local debugging) and only run the whole model CI.

File: test/srt/models/test_nvidia_nemotron_nano_v2.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            data_path=None,
+            num_questions=200,
+            max_new_tokens=512,
+            parallel=128,
+            host="http://127.0.0.1",
+            port=int(self.base_url.split(":")[-1]),
+        )
+        metrics = run_eval(args)
+        print(f"{metrics=}")
+        self.assertGreater(metrics["accuracy"], 0.87)
```
Comment (merrymercy) [inline test/srt/models/test_nvidia_nemotron_nano_v2.py:44]: We need to add this. Otherwise, it won't run by the CI.

https://github.com/sgl-project/sglang/blob/55b14656e656c0a0be3d9a24e69f624c957429e7/test/srt/test_abort.py#L113-L114

File: python/sglang/srt/layers/attention/attention_registry.py [hunk 9]
```diff
@@ -166,36 +173,41 @@ def create_dual_chunk_flash_attn_backend(runner):
...
 
-    # wrap for hybrid GDN models
-    if runner.is_hybrid_gdn:
-        from sglang.srt.utils import is_blackwell, is_npu
-
-        if is_blackwell():
-            assert (
-                runner.server_args.attention_backend == "triton"
-                or runner.server_args.attention_backend == "trtllm_mha"
-            ), "triton or trtllm_mha backend are the only supported backends on Blackwell GPUs for hybrid GDN models, use --attention-backend triton or --attention-backend trtllm_mha to specify the backend."
```
Comment (samuellees) [inline python/sglang/srt/layers/attention/attention_registry.py:186]: Here disable trtllm_mha
